diff -uNr w32api-3.9/include/mprapi.h w32api-3.9-xorp/include/mprapi.h
--- w32api-3.9/include/mprapi.h	2006-05-12 18:12:31.000000000 +0100
+++ w32api-3.9-xorp/include/mprapi.h	2007-05-29 21:45:53.265625000 +0100
@@ -187,6 +187,8 @@
 	GUID guid;
 	PPP_INFO_2 PppInfo2;
 } RAS_CONNECTION_2,*PRAS_CONNECTION_2;
+typedef void *RAS_CONNECTION_3;
+typedef RAS_CONNECTION_3* PRAS_CONNECTION_3;
 #define MAX_PORT_NAME 16
 #define MAX_MEDIA_NAME 16
 #define MAX_DEVICE_NAME 128
diff -uNr w32api-3.9/include/routprot.h w32api-3.9-xorp/include/routprot.h
--- w32api-3.9/include/routprot.h	2006-05-11 16:57:11.000000000 +0100
+++ w32api-3.9-xorp/include/routprot.h	2007-05-29 21:46:07.671875000 +0100
@@ -43,6 +43,10 @@
 #define IPX_PROTOCOL_NLSP 0x00020002
 /*--- Router Management Reference - Router Management Structures */
 #if (_WIN32_WINNT >= 0x0500)
+typedef struct IP_LOCAL_BINDING {
+	DWORD Address;
+	DWORD Mask;
+} IP_LOCAL_BINDING,*PIP_LOCAL_BINDING;
 typedef struct IP_ADAPTER_BINDING_INFO {
 	ULONG AddressCount;
 	DWORD RemoteAddress;
@@ -50,10 +54,6 @@
 	ULONGLONG Speed;
 	IP_LOCAL_BINDING Address[];
 } IP_ADAPTER_BINDING_INFO,*PIP_ADAPTER_BINDING_INFO;
-typedef struct IP_LOCAL_BINDING {
-	DWORD Address;
-	DWORD Mask;
-} IP_LOCAL_BINDING,*PIP_LOCAL_BINDING;
 typedef struct IPX_ADAPTER_BINDING_INFO {
 	ULONG AdapterIndex;
 	UCHAR Network[4];
