--- unistd.h.orig	Thu Jan  5 13:44:02 2006
+++ unistd.h	Sat Jan  7 10:05:59 2006
@@ -10,24 +10,4 @@
 #include <io.h>
 #include <process.h>
 
-#define __UNISTD_GETOPT__
-#include <getopt.h>
-#undef __UNISTD_GETOPT__
-
-#ifdef __cplusplus
-extern "C" {
-#endif
-
-/* This is defined as a real library function to allow autoconf
-   to verify its existence. */
-int ftruncate(int, off_t);
-__CRT_INLINE int ftruncate(int __fd, off_t __length)
-{
-  return _chsize (__fd, __length);
-}
-
-#ifdef __cplusplus
-}
-#endif
-
 #endif /* _UNISTD_H */
